### ISSUE TITLE
Run Gemma 2: unigram tokenization, add_space_prefix, an independent d_head, two extra norms, sliding-window attention and softcapping

### ISSUE THREAD

**Issue by xoloki** [xoloki/jlib#174]
Gemma 2 fails before the architecture gate is even reached, in the tokenizer:

```
$ jchat gemma-2-2b-it-Q4_K_M.gguf
tokenizer: the file carries no merge list, and the scores in these files are
           not usable -- see tokenizer.hh
```

Measured from `gemma-2-2b-it-Q4_K_M.gguf`. Unlike Qwen 2.5 (#173), which needs three small things, this is **five independent defects**, and one of them is a different tokenization algorithm rather than a variant of the existing one. It is an arc, not a patch.

Its chat template already renders -- that was done on the `ai-chat-templates` branch and is checked against python-jinja2.

## 1. A llama vocabulary with no merges: unigram, not BPE

`tokenizer.ggml.model` is `llama`, so the flavour reading is right and the byte-level work is irrelevant here. But the file carries **no merge list at all**. It is SentencePiece **unigram**: every piece has a score, and tokenization is a Viterbi search for the highest-scoring segmentation of the text.

`ai::tokenizer` is merge-driven throughout -- `merge_run()` repeatedly finds the best-ranked adjacent pair. There is no way to reach unigram from there by adjusting it; it is a second algorithm alongside the first.

This is the item that makes the issue large, and it is worth stating that the axes are now **three**, not two: vocabulary flavour, tokenization algorithm, and architecture, all independent.

| file | architecture | vocabulary | algorithm |
| --- | --- | --- | --- |
| TinyLlama | llama | llama | merges |
| Llama 3.2 | llama | gpt2 | merges |
| Qwen 2.5 | qwen2 | gpt2 | merges |
| Gemma 2 | gemma2 | llama | **scores** |

## 2. `add_space_prefix = 0`, and nothing reads it

```
tokenizer.ggml.add_space_prefix = 0
```

`grep -rn add_space_prefix jlib/` finds nothing. The sentencepiece path adds the dummy prefix unconditionally. Gemma says not to, so even with unigram implemented, every input would gain a leading space that the model was not trained to see.

Cheap to fix and easy to miss, because it does not fail -- it just shifts the first token of every prompt.

## 3. `d_head` is not `d_model / heads`

```
gemma2.embedding_length     = 2304
gemma2.attention.head_count = 8       -> 2304 / 8 = 288
gemma2.attention.key_length   = 256
gemma2.attention.value_length = 256
```

`transformer.hh:323` computes `m_d_head(heads ? d_model / heads : 0)`. That is 288 here, and the file says 256. The head dimension is an independent quantity that Llama files happen to make derivable, and jlib has taken the coincidence for the rule.

Worth fixing on its own merits even if Gemma is never supported, since it is a silent wrong answer rather than a refusal.

## 4. Two extra norms per block

```
blk.0.post_attention_norm.weight
blk.0.post_ffw_norm.weight
```

Gemma 2 normalises after the attention and feed-forward sublayers as well as before. `transformer.hh` has one norm on each side. Nothing loads these tensors.

## 5. Sliding-window attention and logit softcapping

```
gemma2.attention.sliding_window   = 4096
gemma2.attn_logit_softcapping     (present)
gemma2.final_logit_softcapping    (present)
```

Gemma 2 alternates local (windowed) and global attention by layer, and caps attention and final logits through a tanh. `ai::attention` is full-attention only, with no softcap. Ignoring these does not crash and does not obviously break short prompts, which is exactly why it needs to be on the list rather than discovered later.

## Suggested order

1. `d_head` from the file (#3) -- standalone, fixes a latent assumption
2. `add_space_prefix` (#2) -- standalone, one key
3. Unigram tokenization (#1), against reference tokenizations from HuggingFace `tokenizers` reading Gemma's own `tokenizer.json` -- not against this library's output
4. The `gemma2` architecture: metadata prefix (shared with #173), the two norms, `d_head`, sliding window, softcapping

1 through 3 are useful whether or not 4 ever happens.

## What would say it works

Reference tokenizations from `tokenizers`, and the model answering a question. As with Qwen, neither alone -- and for the architecture in particular, "it produces fluent text" is not evidence, since every one of defects 3 through 5 produces fluent text.

**Comment by xoloki** [comment]
**Correcting item 1 before implementing it: this is not unigram.**

The issue said Gemma needs "SentencePiece **unigram**, scored rather than merged" and a Viterbi search. Measured, it does not.

## What the file actually carries

```
TinyLlama   32000 tokens,  61249 merges,  all 32000 scores are zero
Gemma 2    256000 tokens,  no merges,     255743 non-zero scores
```

Exact complements. And Gemma's scores are **ranks, not log-probabilities**:

```
index    100 -> -1000     (a special)
index    300 ->     0     (a byte token)
index   1000 ->  -527
index   5000 -> -4527
index  50000 -> -49527
index 128000 -> -127527
```

That is `-(index - 473)` across the ordinary vocabulary, the offset being the specials and byte tokens ahead of it. So the highest-scoring pair is simply the earliest one in the vocabulary.

Which means it is **the same greedy pairwise merge loop already implemented**, with the vocabulary index standing in for the merge rank — SentencePiece's rule, take whichever adjacent pair makes the highest-scoring token. Not a Viterbi search over segmentations, and not a second algorithm.

Corroborated from the other side: Gemma's own `tokenizer.json` is `"type": "BPE"` with 580,604 merges. The GGUF conversion dropped the merge list and kept the scores; HF's kept the merges. Both describe the same tokenizer.

## What it took

Two changes, not an arc:

- `merge_key()` — the one place that knows which signal drives the loop. Under merges the key is the pair's rank; under scores it is the negated score of the token the pair would make, so the caller compares lower-is-better either way.
- `add_space_prefix` (item 2) — Gemma sets it to 0, and `decode()` had to follow, or it would strip a space the text really had.

**28 of 28 reference tokenizations match**, from HuggingFace `tokenizers` reading Gemma's own tokenizer.json. TinyLlama, Llama 3.2 and Qwen 2.5 are unchanged.

The driver is a property of the **file**, not of the flavour — Gemma and TinyLlama are both `tokenizer.ggml.model == "llama"` and are driven the opposite ways. Inferring it from the flavour would get both wrong, which is the same shape as the vocabulary/architecture lesson from #177.

Items 3, 4 and 5 — `d_head`, the two extra norms, sliding window and softcapping — are still ahead and are architecture rather than tokenizer.

**Comment by xoloki** [comment]
Architecture half implemented. **Gemma does not run yet, and the reason is not what the issue predicted.**

## Done and verified

- **`d_head` from the file** (item 3). `attention.key_length` is 256 where 2304/8 is 288, and the tensor shapes confirm it independently: `attn_q` is 2304x**2048** = 8 x 256, `attn_k` 2304x**1024** = 4 x 256. `load()` derived it the same wrong way and would have checked shapes the file does not have. llama and qwen2 verified unchanged.
- **The two post-sublayer norms** (item 4), applied to the sublayer output before the residual, as `modeling_gemma2` does.
- **Softcapping** (item 5, half): a new `softcap` backend op on the host and in Metal, attention logits at 50 before the mask -- before, because the mask writes -infinity and capping that would let a masked position back in -- and final logits at 30.

## Two the issue does not list, both load-bearing

- **The embedding is scaled by sqrt(d_model)** = 48. Applied after the gather and *not* to the table, because Gemma ties its head to the embedding and scaling the table would scale the logits too.
- **RMSNorm weights mean `1 + w`.** Applied once at load rather than in the kernel: it is a property of the stored numbers, not of the operation, and a branch there would cost every architecture for one.

## And a third: Gemma is GeGLU, not SwiGLU

`activation::gelu` did not exist. Added to both backends in the tanh form -- `gelu_pytorch_tanh`, which is what Gemma was trained with; the erf form differs in the fourth decimal and a model is entitled to the one it saw.

## The open problem: Metal diverges from the host at four layers

```
                    2 layers    3 layers    4 layers
host   float        29.9849     29.7964     29.896
metal  float        29.9849     29.7964     256000 NaN
metal  fp16         29.9844     29.7969     256000 NaN
```

All three agree to four decimals at two and three layers. At four, **both** Metal element types produce an all-NaN logit vector and the host does not.

So it is neither a precision problem nor the architecture logic:

- fp16 range was the obvious hypothesis and is **wrong** -- `metal::backend<float>` fails identically.
- The host in float produces finite, plausible logits at the same depth with the same weights.
- Not the softcapping: NaN with the caps on, off, and each alone.
- Not aliasing in the post-norm: `k_rms_norm` reads a whole column before writing and one thread owns a column.
- `k_attn_scores` has no fixed-size assumption -- it loops over `d_head` as a parameter and accumulates in float.

The one thing that is unlike every model Metal handles today is **`d_head` = 256, where llama and qwen2 are 64**. That is where I would look next: a Metal kernel whose indexing or dispatch is fine at 64 and wrong at 256, exposed only now. It would be a pre-existing bug rather than one this branch introduced.

Sliding window (4096, item 5's other half) is not implemented and is not the cause -- the probes above are 15-token prompts.

TinyLlama, Llama 3.2 and Qwen 2.5 all still answer correctly; suite is 72 pass, 4 skip, 0 fail.

**Comment by xoloki** [comment]
Done in #180 — Gemma 2 runs:

```
The capital of France is **Paris**. 🇫🇷
```

28/28 reference tokenizations match. The remaining item, **sliding-window attention**, is split out as #181: it is the identity below 4096 tokens, which is why Gemma answers correctly without it.

## This issue got two of its five items wrong, which is the useful part

**Item 1 was not unigram.** It said Gemma needed a Viterbi search, "a second algorithm alongside the first". Measured, Gemma and TinyLlama are exact complements — 61249 merges and all-zero scores against no merges and 256000 real ones — and Gemma's scores are *ranks*, `-(index - 473)`, so the best-scoring pair is the earliest in the vocabulary. It is the greedy merge loop already implemented with the index standing in for the rank. Corroborated from the other side: Gemma's own `tokenizer.json` is BPE with 580604 merges; the GGUF conversion dropped them and kept the scores.

**The list was missing three conventions**, all load-bearing and none in the file: the embedding scaled by sqrt(d_model), GeGLU rather than SwiGLU (`activation::gelu` did not exist in either backend), and the split RoPE layout.

## And two things found only by running it

**Metal's `tanh` returns NaN past x≈44** — `(exp(2x)-1)/(exp(2x)+1)` overflowing, and only for positive x, which is the asymmetry that identified it. **Pre-existing**: `activation::tanh` had it too, 1 at x=20 and NaN at x=128. Nothing had fed it a large enough argument before GeGLU. I nearly concluded "Gemma needs bf16" instead — `metal::backend<float>` failing identically is what killed that, and the op-by-op host comparison localised it.

**The norm offset was already applied.** Gemma's published implementation stores RMSNorm weights as `1 + w`, so I did that, and the model answered "1" to everything. The GGUF conversion had already done it: `attn_norm` runs [0.000, 5.969] mean 1.19, where TinyLlama's raw weights run [-0.582, 0.770]. **The HF checkpoint and the GGUF are different artefacts**, and a convention read out of `modeling_gemma2.py` describes the former.

Also discovered and filed separately: jlib cannot read K-quantised files at all, which is how most GGUFs ship.
